# Assignment 06: Text Analysis

## Step 1: Make a new Jupyter notebook

In [68]:
import requests
from bs4 import BeautifulSoup
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
from string import punctuation
import nltk
from nltk.sentiment import vader
nltk.download('vader_lexicon')
import pandas as pd
import matplotlib as plt
from nltk.stem.porter import PorterStemmer
from gensim import corpora, models
import gensim


[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


## Step 2: Shakespeare word frequency

Make a Python string that contains the text of a Shakespeare play (obtained, for example, from Project Gutenberg)

In [3]:
response = requests.get('https://www.gutenberg.org/files/1524/1524-h/1524-h.htm')
html_string = response.text
document = BeautifulSoup(html_string, "html.parser")

In [4]:
document = document.text

Tokenize the words and remove stopwords

In [6]:
for i in ['\n','[',']','’','”','“', '\r']:
    document = document.replace(i,' ')

In [7]:
print(document)

     The Project Gutenberg eBook of Hamlet, by William Shakespeare     The Project Gutenberg eBook of Hamlet, by William Shakespeare   This eBook is for the use of anyone anywhere in the United States and  most other parts of the world at no cost and with almost no restrictions  whatsoever. You may copy it, give it away or re-use it under the terms  of the Project Gutenberg License included with this eBook or online  at www.gutenberg.org. If you  are not located in the United States, you will have to check the laws of the  country where you are located before using this eBook.   Title: Hamlet Author: William Shakespeare Release Date: November 1998  eBook #1524    Most recently updated: October 17, 2021  Language: English Character set encoding: UTF-8 Produced by: Dianne Bean *** START OF THE PROJECT GUTENBERG EBOOK HAMLET ***    THE TRAGEDY OF HAMLET, PRINCE OF DENMARK by William Shakespeare  Contents   ACT I Scene I.  Elsinore. A platform before the Castle. Scene II.  Elsinore. A room

In [9]:
sent = sent_tokenize(document)
word = word_tokenize(document.lower())

In [10]:
print(word)

['the', 'project', 'gutenberg', 'ebook', 'of', 'hamlet', ',', 'by', 'william', 'shakespeare', 'the', 'project', 'gutenberg', 'ebook', 'of', 'hamlet', ',', 'by', 'william', 'shakespeare', 'this', 'ebook', 'is', 'for', 'the', 'use', 'of', 'anyone', 'anywhere', 'in', 'the', 'united', 'states', 'and', 'most', 'other', 'parts', 'of', 'the', 'world', 'at', 'no', 'cost', 'and', 'with', 'almost', 'no', 'restrictions', 'whatsoever', '.', 'you', 'may', 'copy', 'it', ',', 'give', 'it', 'away', 'or', 're-use', 'it', 'under', 'the', 'terms', 'of', 'the', 'project', 'gutenberg', 'license', 'included', 'with', 'this', 'ebook', 'or', 'online', 'at', 'www.gutenberg.org', '.', 'if', 'you', 'are', 'not', 'located', 'in', 'the', 'united', 'states', ',', 'you', 'will', 'have', 'to', 'check', 'the', 'laws', 'of', 'the', 'country', 'where', 'you', 'are', 'located', 'before', 'using', 'this', 'ebook', '.', 'title', ':', 'hamlet', 'author', ':', 'william', 'shakespeare', 'release', 'date', ':', 'november', '19

In [14]:
myStopWords = list(punctuation) + stopwords.words('english')
wordsNoStopWords = [w for w in word if w not in myStopWords]

In [15]:
wordsNoStopWords

['project',
 'gutenberg',
 'ebook',
 'hamlet',
 'william',
 'shakespeare',
 'project',
 'gutenberg',
 'ebook',
 'hamlet',
 'william',
 'shakespeare',
 'ebook',
 'use',
 'anyone',
 'anywhere',
 'united',
 'states',
 'parts',
 'world',
 'cost',
 'almost',
 'restrictions',
 'whatsoever',
 'may',
 'copy',
 'give',
 'away',
 're-use',
 'terms',
 'project',
 'gutenberg',
 'license',
 'included',
 'ebook',
 'online',
 'www.gutenberg.org',
 'located',
 'united',
 'states',
 'check',
 'laws',
 'country',
 'located',
 'using',
 'ebook',
 'title',
 'hamlet',
 'author',
 'william',
 'shakespeare',
 'release',
 'date',
 'november',
 '1998',
 'ebook',
 '1524',
 'recently',
 'updated',
 'october',
 '17',
 '2021',
 'language',
 'english',
 'character',
 'set',
 'encoding',
 'utf-8',
 'produced',
 'dianne',
 'bean',
 'start',
 'project',
 'gutenberg',
 'ebook',
 'hamlet',
 'tragedy',
 'hamlet',
 'prince',
 'denmark',
 'william',
 'shakespeare',
 'contents',
 'act',
 'scene',
 'i.',
 'elsinore',
 'platf

Find the top 20 most frequent words in the play

In [16]:
from nltk.probability import FreqDist

In [19]:
twenty = FreqDist(wordsNoStopWords)

In [20]:
twenty

FreqDist({'â': 494, 'hamlet': 479, 'lord': 222, 'king': 205, 'horatio': 156, 'polonius': 125, 'queen': 120, 'shall': 116, 'laertes': 108, 'good': 107, ...})

In [21]:
for i in sorted(twenty, key=twenty.get, reverse=True)[:20]:
    print(i,freq[i])

â 494
hamlet 479
lord 222
king 205
horatio 156
polonius 125
queen 120
shall 116
laertes 108
good 107
come 106
thou 105
let 104
project 90
ophelia 87
may 86
thy 86
would 81
like 80
know 79


Comment on whether these words give an accurate sense of the play

I think that the top twenty words do not give an accurate sense of the play. A majority of the words are the names of the main characters, and the names likely appear a lot in the text because it is a script. Therefore, this likely gives a better picture of which characters have more lines in the play. I think to get an accurate sense of the themes and contents of the play, there needs to be more data cleaning removing words from the website like 'project' and character names like 'ophelia'.

## Step 3: Yelp sentiments

Find your favorite restaurant on Yelp and copy 15 of its reviews into your notebook as Python strings

In [44]:
# Reviews for Tsujita LA
polarity_scores = []
stars = []

In [24]:
sia = vader.SentimentIntensityAnalyzer()

Use Vader to find the polarity of each review

In [27]:
review1 = "Yum yum! I went here on a weekend night I believe around 630pm. When I walked in with my friend, the place was already pretty crowded as people were waiting to get seated and there were a few parties written on the waitlist before me. However, my friend and I actually ended up getting seated in a span of around 5 min since there was an outside seat for 2 available for us!I ended up ordering the spicy tonkotsu ramen with an egg added. I also got the noodles extra hard since I prefer it that way.  Of course, I started the meal with sipping on the broth first. There was so much tonkotsu flavor, very satisfied! The chasiu was decently soft too, and the noodles perfectly cooked. Just one thing though, is that the egg was very well cooked but there wasn't much flavor :/ a little disappointingOne other thing, is just that even though I was seated outside, I could smell the tonkotsu and it was rather strong and unpleasant.With all that being said, I still give them 5 stars for my overall experience. The waiters were extremely helpful and nice as well. I will come back here sometime for their Tsukemen next time, since I heard those are yummy!"

stars.append(5)
polarity_scores.append(sia.polarity_scores(review1)['compound'])


In [28]:
review2 = "Love the food, hate the wait Make sure to ask for broth for the thick ramen dipping sauce when you're 70% done with the noodles!"

stars.append(4)
polarity_scores.append(sia.polarity_scores(review2)['compound'])


In [29]:
review3 = "It's very good ramen, the tsukemen dip style is delicious. The noodles are thick and chewy, and the dipping sauce is rich and tasty. There can be quite a wait, though the line goes relatively quick."

stars.append(4)
polarity_scores.append(sia.polarity_scores(review3)['compound'])


In [30]:
review4 = "Service was great, and the food came out quick. Would highly recommend."

stars.append(4)
polarity_scores.append(sia.polarity_scores(review4)['compound'])


In [31]:
review5 = "Decent ramen, great service, tough location. This place is on Sawtelle so parking is impossible and you'll definitely have to wait. I got the charsiu ramen with medium spice and I definitely wish I had stuck with the tsukkemen, although that dish isn't my favorite."

stars.append(3)
polarity_scores.append(sia.polarity_scores(review5)['compound'])


In [32]:
review6 = "Overall, the ramen was very good. For its pricing, I'd say it was worth paying for. The only reason I took off one star is because the restaurant has some kind of odd pungent smell, which bothered me a lot while I was dining."

stars.append(4)
polarity_scores.append(sia.polarity_scores(review6)['compound'])


In [33]:
review7 = "Service was friendly and efficient. We liked the Charsiu Tsukemen and had enough leftovers to take some home although it for sure is best fresh!"

stars.append(4)
polarity_scores.append(sia.polarity_scores(review7)['compound'])


In [34]:
review8 = "The best spot in LA for Tsukemen! I normally get the charsiu tsukemen ramen but I realized I could never finish the char siu."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review8)['compound'])


In [35]:
review9 = "Hands down still the best tsukemen I've ever had. I always recommend people to try Tsujita for their dipping ramen because it's always just so fresh and flavourful. The egg is also always so perfectly made."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review9)['compound'])


In [36]:
review10 = "Really good in person, slightly less great to go. Still the king of tsukemen in LA. Undisputedly still the best tsukemen you can find in LA, especially since the death of Chinatown's Okiburo. The dipping broth which is more like a gravy is thick, almost like watered down curry. The flavor is intense and full of porky and fishy flavor, in the best way possible."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review10)['compound'])


In [37]:
review11 = "I always go here when I'm in LA! It's one of the best I've ever had."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review11)['compound'])


In [38]:
review12 = "Always have seen a line here, but ya'll it's worth the wait. The dip ramen was amazing. Chewy thick noodles and super rich dipping broth."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review12)['compound'])


In [39]:
review13 = "I'd say this place is worth a try but I wouldn't recommend getting the charsiu. There's definitely better out there."

stars.append(4)
polarity_scores.append(sia.polarity_scores(review13)['compound'])


In [40]:
review14 = "This place was highly recommended to me from multiple people and I can see why. I've never had tsukemen before but it was fun and delicious. I heard from others that the wait would be crazy but I came on a Sunday night at 7 and the wait was less than 10 min."

stars.append(5)
polarity_scores.append(sia.polarity_scores(review14)['compound'])


In [41]:
review15 = "I'm gonna be honest, this might have been the best ramen I've ever had. Tsujita is a must-try place if you are in LA. 100% would recommend :)"

stars.append(5)
polarity_scores.append(sia.polarity_scores(review15)['compound'])


## Compare Vader's scores against user-specified numbers of stars

In [43]:
pd.DataFrame({'Stars': stars,
              'Vader Score': polarity_scores},
                columns = ['Stars', 'Vader Score'])

Stars  Vader Score
0       5       0.9886
1       4       0.4753
2       5       0.9886
3       4       0.4753
4       4       0.8883
5       4       0.7841
6       3       0.8031
7       4       0.0516
8       4       0.9508
9       5       0.4389
10      5       0.9438
11      5       0.9505
12      5       0.6696
13      5       0.9628
14      4       0.7339
15      5       0.8379
16      5       0.9186

From the dataframe above, the results comparing the Yelp stars with the Vader polarity compound scores of the review are unexpected. Although most of the reviews are 4 stars and above, there are still many reviews with a low Vader score especially review 7 (0.0516 Vader score). This is likely because most of the reviews mention the long wait times in Tsujita which skews the Vader polarity score to detect negative connotations.

## Step 4: Your movie reviews

Make 5 strings that contain reviews (3 sentences each) of your favorite movie comedies

In [47]:
shrek = "Turning the fairy-tale genre on its head was a clever, if not totally novel, notion at the time, and Shrek still retains much of its ironic charm 20 years later. With the perfect blend of storytelling, animation, and music this is one movie (Shrek) the whole family can enjoy! Shrek has a positive message about accepting yourself for who you are."


In [48]:
shrek2 = "Introduces one of the most lovable and laughable characters in animation history: Puss in Boots. It's just as good - if not better - than the original. The animation by which it stands or falls is as brilliant as ever and, though it wouldn't really be right to call it totally anti-Disney, it certainly trumps that institution for sharpness of focus, notably as far as the screenplay is concerned."


In [49]:
catinthehat = "Filled with double entendres and potty humor, this Cat falls flat. Part of the problem is the film confuses chaos with fun. This film was visually appealing yet devoid of any substance."


In [50]:
crazystupidlove = "A rare romantic comedy with substance and heart. Surprisingly good, especially for a romantic-comedy, as the bar is set very low in this genre. It's more the comedy than the romance that works though. 'Crazy, Stupid, Love.' is a fun film to watch. It's a well-made, entertaining & superbly acted romantic-comedy, that earns itself a viewing due to it's true merits."


In [51]:
easya = "It owes a huge debt to older (and better) teen comedies, but Easy A proves a smart, witty showcase for its irresistibly charming star, Emma Stone. Arguably the best teen comedy since Clueless, it's easy to give this one an A. Well, A-. With scintillating wit and endlessly quotable dialogue, Easy A is a potent critique of retrograde attitudes towards female sexuality."


Make 5 strings that contain reviews (3 sentence each) of your favorite movie dramas

In [52]:
parasite = "An urgent, brilliantly layered look at timely social themes, Parasite finds writer-director Bong Joon Ho in near-total command of his craft. Bong Joon-ho has created something with Parasite thats darkly humorous, compelling, dramatic, poignant, and bittersweet all at the same time. A bonkers, beautiful, radical, and drop-dead intelligent dark satirical tale of social inequality [and the] mock egalitarian weirdness of late capitalism."


In [54]:
hungergames = "Thrilling and superbly acted, The Hunger Games captures the dramatic violence, raw emotion, and ambitious scope of its source novel. The Hunger Games is aptly lensed, shrewdly cast, and keen to make the most of its glaring cultural parallels. Overall, The Hunger Games offers a thrilling experience with spectacular and innovative elements and views."


In [55]:
twilight = "Having lost much of its bite transitioning to the big screen, Twilight will please its devoted fans, but do little for the uninitiated. I've had mosquito bites that were more passionate than this undead, unrequited, and altogether unfun pseudo-romantic riff on Romeo and Juliet. Twilight stands as a powerful, darkly stylish depiction of teen female desire."


In [56]:
knivesout = "Knives Out sharpens old murder-mystery tropes with a keenly assembled suspense outing that makes brilliant use of writer-director Rian Johnson's stellar ensemble. Knives Out is a thoroughly satisfying murder mystery populated with well-drawn, idiosyncratic characters who collectively present a tapestry of modern America. Funny and smart, this is one of the wittiest and most entertaining films of the year."


In [57]:
ladybird = "Lady Bird delivers fresh insights about the turmoil of adolescence -- and reveals writer-director Greta Gerwig as a fully formed filmmaking talent. Saoirse Ronan's stunning portrayal of adolescence is completely lacking in vanity. With Lady Bird, Greta Gerwig manages to take the old coming-of-age genre and make it feel refreshing with its authenticity and heart."


Make a Python list that contains these 10 strings

In [58]:
reviews = [shrek, shrek2, catinthehat,
           crazystupidlove, easya, parasite,
           hungergames, twilight, knivesout,
           ladybird]

Replicate the analysis pipeline from "04_news_topics.ipynb"

In [60]:
extrastop = ['``',"''","'re","'s","'re",'``',"''","'ll","--","\'\'","...",
             "n\'t",'one','would','use','subject','from',"\'m","\'ve"]

myStopWords = list(punctuation) + stopwords.words('english') + extrastop


In [63]:
[w for w in word_tokenize(reviews[0].lower()) if w not in myStopWords]

reviewswords = []
for i in reviews:
    reviewswords.append([w for w in word_tokenize(i.lower()) if w not in myStopWords])


In [66]:
p_stemmer = PorterStemmer()

listOfStemmedWords = []
for i in reviewswords:
    listOfStemmedWords.append([p_stemmer.stem(w) for w in i])

In [67]:
!pip install gensim

     |████████████████████████████████| 24.0 MB 17.9 MB/s eta 0:00:01
     |████████████████████████████████| 58 kB 1.9 MB/s  eta 0:00:01


In [69]:
dictionary = corpora.Dictionary(listOfStemmedWords)
corpus = [dictionary.doc2bow(text) for text in listOfStemmedWords]
ldamodel = gensim.models.ldamodel.LdaModel(corpus, 
                                           num_topics=2, 
                                           id2word = dictionary, 
                                           passes=20)


In [70]:
for i in ldamodel.print_topics(num_topics=2, num_words=20):
    print(i)

(0, '0.015*"comedi" + 0.012*"film" + 0.012*"easi" + 0.012*"anim" + 0.012*"shrek" + 0.008*"teen" + 0.008*"genr" + 0.008*"better" + 0.008*"good" + 0.008*"though" + 0.008*"romantic-comedi" + 0.008*"total" + 0.008*"charm" + 0.008*"fun" + 0.008*"substanc" + 0.008*"fall" + 0.005*"femal" + 0.005*"smart" + 0.005*"stand" + 0.005*"charact"')
(1, '0.011*"game" + 0.011*"writer-director" + 0.011*"hunger" + 0.011*"make" + 0.008*"time" + 0.008*"social" + 0.008*"bong" + 0.008*"parasit" + 0.008*"dramat" + 0.008*"darkli" + 0.008*"knive" + 0.008*"old" + 0.008*"thrill" + 0.008*"twilight" + 0.008*"greta" + 0.008*"bird" + 0.008*"bite" + 0.008*"ladi" + 0.008*"gerwig" + 0.008*"adolesc"')


Comment on the words that the model chooses to represent the 2 topics, and whether they match with your split between comedies and dramas

The words chosen for the comedies match the themes better since the model selected words like "comedic". However, the words chosen for the dramas are contain a lot of names of the directors or actors like "greta gerwig" and "bong jo ho". As a surface analysis, this may be indicative of how drama movies are deemed as more critically acclaimed and thus considered a merit to the careers of filmmakers and actors.